File: ovp8xx/cpp/core/fw_update/fw_update.cpp
/*
 * Copyright 2024-present ifm electronic, gmbh
 * SPDX-License-Identifier: Apache-2.0
 */
// This example shows how to update the firmware
// of an O3R device.

#include <filesystem>
#include <ifm3d/device/o3r.h>
#include <ifm3d/swupdater/swupdater.h>
#include <iostream>

using namespace ifm3d::literals;

int main() {
  // Provide the path to the sofware update file:
  std::filesystem::path FILENAME =
      "/path/to/o3r/fw/OVP81x_Firmware_1.10.13.5502.swu";
  if (std::filesystem::exists(FILENAME)) {
    std::cout << "File exists: " << FILENAME << std::endl;
  } else {
    std::cerr << "File does not exist: " << FILENAME << std::endl;
  }

  // Getting the IP address from the environment variable or
  // using the default IP address if not defined.
  const char *IP = std::getenv("IFM3D_IP") ? std::getenv("IFM3D_IP")
                                           : ifm3d::DEFAULT_IP.c_str();
  auto o3r = std::make_shared<ifm3d::O3R>(IP);
  auto swu = std::make_shared<ifm3d::SWUpdater>(o3r);

  // Reboot the system to recovery mode
  std::cout << "Rebooting to recovery mode..." << std::endl;
  swu->RebootToRecovery();
  swu->WaitForRecovery();

  // Flash the firmware
  std::cout << "Flashing firmware..." << std::endl;
  if (swu->FlashFirmware(FILENAME, 1800)) {
    swu->WaitForProductive();
    std::cout << "Firmware update successful. System ready!" << std::endl;
  } else {
    std::cerr << "Firmware update failed." << std::endl;
  }

  // Check the version after update
  std::cout << "Current version: " << o3r->Get({"/device/swVersion/firmware"})
            << std::endl;

  return 0;
}

File: ovp8xx/python/core/fw_update.py
# -*- coding: utf-8 -*-
################################################
# Copyright 2025-present ifm electronic, gmbh
# SPDX-License-Identifier: Apache-2.0
################################################
# This examples shows how to update the firmware
# of an O3R device.
################################################

from os import path

from ifm3dpy.device import O3R
from ifm3dpy.swupdater import SWUpdater

###################################
# Configure the IP and the path to
# the firmware file on your system
###################################
IP = "192.168.0.69"
FILENAME = "/path/to/o3r/fw/OVP81x_Firmware_1.20.29.6418.swu"
o3r = O3R(IP)
swu = SWUpdater(o3r)

if not path.exists(FILENAME):
    raise ValueError("Provided swu file does not exist")

try:
    print("Rebooting to recovery mode...")
    swu.reboot_to_recovery()
    swu.wait_for_recovery()
    if swu.flash_firmware(swu_file=FILENAME, timeout_millis=1800):
        swu.wait_for_productive()
        print("Update successful. System ready!")
    else:
        raise TimeoutError("Timeout during firmware update")
except RuntimeError:
    raise RuntimeError("Firmware update failed!")

print("Current firmware version: ")
print(o3r.get(["/device/swVersion/firmware"]))
